Deprecates indexing and querying a context completion field without context

jimczi opened the pull request:
This change deprecates completion queries and documents without context that target a
context enabled completion field. Querying without context degrades the search
performance considerably (even when the number of indexed contexts is low).
This commit targets master but the deprecation will take place in 6.x and the functionality
will be removed in 7 in a follow up.

Closes #29222

jimczi committed:
Deprecates completion query without context

This change deprecates completion queries without context that target a
context enabled completion field. Querying without context degrades the search
performance considerably (even when the number of indexed contexts is low).
This commit targets master but the deprecation will take place in 6.x and the functionality
will be removed in 7 in a follow up.

Closes #29222

elasticmachine commented:
Pinging @elastic/es-search-aggs

jimczi committed:
fix rest tests

cbuescher commented on server/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggestionBuilder.java:
```diff
@@ -298,16 +304,19 @@ public SuggestionContext build(QueryShardContext context) throws IOException {
+        CompletionFieldMapper.CompletionFieldType type = (CompletionFieldMapper.CompletionFieldType) mappedFieldType;
+        suggestionContext.setFieldType(type);
+        if (type.hasContextMappings()) {
+            if (contextBytes == null) {
```
This checks for a missing `contexts` element in the query, but I think it will miss cases like `"contexts" : {}` or `"contexts" :  { "someField : [] }` which should also emit a deprecation warning. Not sure if this means there should be another check after the queryContexts map is parsed.

jimczi replied:
Good catch, thanks ! This also made me realize that we should deprecate the index side if no contexts are provided. This is needed because these suggestions are not retrievable anymore if we remove the ability to search with no context. I pushed https://github.com/elastic/elasticsearch/pull/30712/commits/f5edec43302eafc2edfea583b93e0b0996e5dc9a to handle all cases of empty context in the query side and to also deprecate indexing without contexts.

cbuescher reviewed (changes requested):
I left a comment, rest looks good I think. I checked the two cases I mentioned with rest tests, they don't log a warning atm but I think it would be good if they would.

jimczi committed:
After review

This commit moves the deprecation warning in order to ensure that all cases of empty contexts are handled.
This change also deprecates the indexation of a context completion field without contexts. This is needed because these suggestions
are not reachable anymore if querying without context is removed.

jimczi committed:
Merge branch 'master' into context_completion_deprecation

cbuescher reviewed (approved):
Great, LGTM

jimczi committed:
fix rest test

jimczi commented:
This change deprecates completion queries and documents without context that target a
context enabled completion field. Querying without context degrades the search
performance considerably (even when the number of indexed contexts is low).
This commit targets master but the deprecation will take place in 6.x and the functionality
will be removed in 7 in a follow up.

Closes #29222

JustinMoser commented:
Hi! Really sorry to comment on this but ive been looking into why my completion suggester queries are so slow when no contexts are provided, and found this (and linked) threads.

Understanding now why not including contexts on a query for an index that has contexts enabled, i now wish to remove the contexts from the completion mapping. Is this possible without recreating the entire index? Sorry if this is the wrong place to comment this!

Thanks :)

bhavik08 commented:
Is there any possible workaround to search all contexts (even if its slow)... We have a use-case where the query can contain no context at all..
Currently I'm planning to add a new dummy field which remains same for all the documents in the index.. 
(In a way, this slows down the query just like querying without a context would've done because it scans all the documents.. But latency is not that big of a concern for us. Just looking for a cleaner way to scan all documents for query)

jtibshirani commented:
@JustinMoser I'm sorry we missed your question earlier. I don't think it is possible to remove contexts from a mapping after it has already been configured. In general, the [discuss forums](https://discuss.elastic.co/) are a good place to ask this type of question, since we reserve GitHub discussions for bug reports and development discussions.

@bhavik08 as I just mentioned, it'd be best to ask this question in the [discuss forums](https://discuss.elastic.co/). You could give some details on your use case, including an example of what you're trying to accomplish.
